File: TustlerServicesLib/RetainingStack.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;

namespace TustlerServicesLib
{
    /// <summary>
    /// Defines a type whose internal items can be consumed until none are left
    /// </summary>
    public interface IConsumable
    {
        public int Count { get; }
        public int Total { get; }
        public void Consume();      // consume the current item
    }

    /// <summary>
    /// A stack-like object that supports Stack semantics but retains all data
    /// </summary>
    /// <remarks>A retaining stack can be consumed but the original contents are always retrievable</remarks>
    public class RetainingStack<T> : IConsumable, IEnumerable<T>
    {
        /// <summary>
        /// Specifies whether stack items such as tasks are independant or sequentially dependant
        /// </summary>
        public enum ItemOrdering
        {
            Independant,    // execution of each item is independant of that of its peers
            Sequential      // items are executed sequentially with sequential dependancies
        }

        private readonly Stack<T> _stack;
        private readonly ImmutableArray<T> _array;

        public RetainingStack(IEnumerable<T> items) : this(items, ItemOrdering.Sequential)
        {
        }

        public RetainingStack(IEnumerable<T> items, ItemOrdering ordering)
        {
            if (items is null) throw new ArgumentNullException(nameof(items), "Expecting a non-null value");

            _array = items.ToImmutableArray();
            _stack = new Stack<T>(items.Reverse());     // reversed so that calling Pop() removes items in _array order

            Ordering = ordering;
        }

        public ItemOrdering Ordering { get; }

        /// <summary>
        /// Get the current (consumable) count
        /// </summary>
        public int Count
        {
            get
            {
                return _stack.Count;
            }
        }

        /// <summary>
        /// Get the total count
        /// </summary>
        public int Total
        {
            get
            {
                return _array.Length;
            }
        }

        /// <summary>
        /// Get the current item (head of stack) without consuming it
        /// </summary>
        public T Current
        {
            get
            {
                return _stack.Peek();
            }
        }

        /// <summary>
        /// Consume the current item (head of stack) and return it
        /// </summary>
        /// <returns></returns>
        public T Pop()
        {
            return _stack.Pop();
        }

        /// <summary>
        /// Consume the current item (head of stack)
        /// </summary>
        public void Consume()
        {
            _stack.Pop();
        }

        /// <summary>
        /// Refill the stack with the same items used at construction time
        /// </summary>
        public void Reset()
        {
            _stack.Clear();
            foreach (var item in _array.Reverse())
            {
                _stack.Push(item);
            }
        }

        public override string ToString()
        {
            return $"RetainingStack of {typeof(T).Name}: store={_array.Count()}; stack={_stack.Count}";
        }

        public IEnumerator<T> GetEnumerator()
        {
            return ((IEnumerable<T>)_array).GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return ((IEnumerable<T>)_array).GetEnumerator();
        }
    }
}

File: CloudWeaver.Types/RetainingStack.fs
namespace CloudWeaver.Types

open System.Collections
open System.Collections.Generic
open System.Collections.Immutable
open System.Text.Json
open System

type IShareIterationArgument =
    abstract member Serialize : Utf8JsonWriter -> unit
    abstract member ToString: unit -> string

/// Defines a type whose internal items can be consumed until none are left
type IConsumable =
    inherit IEnumerable<IShareIterationArgument>
    abstract member Identifier: Guid with get
    abstract member Total : int with get
    abstract member Remaining : int with get
    abstract member Current : IShareIterationArgument option with get
    abstract member Consume : unit -> unit   // consume the current item
    abstract member Reset : unit -> unit
    abstract member ToString: unit -> string

/// A stack-like object that supports Stack semantics but retains all data
/// A retaining stack can be consumed but the original contents are always retrievable
[<AbstractClass>]
type RetainingStack(uid: Guid, items: IEnumerable<IShareIterationArgument>) =

    do
        if (isNull items) then invalidArg "items" "Expecting a non-null value for items"

    let _array = items.ToImmutableArray()
    let _stack = Stack<IShareIterationArgument>(Seq.rev items)  // reversed so that calling Pop() removes items in _array order

    /// the name of the module that contains the iteration argument type
    abstract ModuleName : string with get

    /// Get the unique identifier for this instance
    member this.Identifier with get() = uid

    /// Get the total count
    member this.Total with get() = _array.Length

    /// Get a count of the remaining (consumable) items
    member this.Remaining with get() = _stack.Count

    /// Get the current item (head of stack) without consuming it
    member this.Current with get() =
        match _stack.TryPeek() with
        | true, item -> Some(item)
        | _ -> None

    /// Consume the current item (head of stack) and return it
    member this.Pop() = _stack.Pop()

    /// Consume the current item (head of stack)
    member this.Consume() = _stack.Pop() |> ignore

    /// Refill the stack with the same items used at construction time
    member this.Reset() =

        _stack.Clear()

        _array
        |> Seq.rev
        |> Seq.iter (fun item -> _stack.Push(item))

    override this.ToString() = sprintf "RetainingStack of IShareIterationArgument: total=%d; remaining=%d" (_array.Length) (_stack.Count)

    interface IConsumable with

        member this.Identifier with get() = this.Identifier

        member this.Total: int = this.Total

        member this.Remaining: int = this.Remaining

        member this.Current: IShareIterationArgument option = this.Current

        member this.Consume(): unit = this.Pop() |> ignore

        member this.Reset(): unit = this.Reset()

        member this.ToString(): string = sprintf "IConsumable: Total %d, Remaining %d, Id %s" this.Total this.Remaining (this.Identifier.ToString())

    interface IEnumerable<IShareIterationArgument> with

        member this.GetEnumerator(): System.Collections.IEnumerator = 
            (_array :> IEnumerable).GetEnumerator()

        member this.GetEnumerator(): IEnumerator<IShareIterationArgument> = 
            (_array :> IEnumerable<IShareIterationArgument>).GetEnumerator()
